### PR TITLE
Preserve SSH config port when field is empty #fixed

### LINKED ISSUE #2513: SSH Port field left empty forces -p 22, overriding Port from ~/.ssh/config (regression)
- Sequel Ace Version (including build number): 5.3.1 (20104)
- Sequel Ace Source (App Store/GitHub/Homebrew): GitHub
- macOS Version: 26.5.2 (25F84)
- Processor Type (Intel/Apple): Apple
- MySQL Version: 8.0.26
- macOS Localization: Japanese

**Is Issue Present in [Latest Beta](https://github.com/Sequel-Ace/Sequel-Ace/releases)?**

Yes. 5.3.1 (20104) is the newest release, and the offending line is still present on `main`.

**Description**

When the **SSH Port** field is left empty, Sequel Ace now passes `-p 22` to `ssh`. This
overrides the `Port` directive resolved from the user's `~/.ssh/config`, so any host alias whose
SSH port is not 22 fails with *"The SSH Tunnel has unexpectedly closed"*.

`HostName` and `User` are unaffected — they still resolve from the config correctly. Only `Port`
is clobbered. That makes the symptom look like "Sequel Ace ignores my ssh config", but the config
*is* being read (Sequel Ace invokes `ssh` with `-F <configfile>`).

This is a regression: it worked in 5.2.0 and earlier.

**Steps To Reproduce**

1. In `~/.ssh/config`:

   ```
   Host myserver
     HostName 10.0.0.5
     User deploy
     Port 2222
   ```

2. Preferences → Network → select `~/.ssh/config` as the SSH config file.
3. New connection → SSH tab:
   - SSH Host: `myserver`
   - SSH User: *(empty)*
   - SSH Port: *(empty)*
4. Connect.

**Expected Behaviour**

The tunnel connects to `10.0.0.5:2222` as `deploy` — i.e. what `ssh -G -F ~/.ssh/config myserver`
reports, and what 5.2.0 did.

**Actual Behaviour**

`ssh` is invoked with `-p 22`, connects to `10.0.0.5:22`, and the tunnel dies with
`The SSH Tunnel has unexpectedly closed.`

**Root Cause**

`Source/Other/Utility/SAConnectionService.swift`

```swift
let sshPort = Int(info.sshPort) ?? 22
```

An empty SSH Port field yields `22` instead of `0`, and `Source/Other/SSHTunnel/SPSSHTunnel.m`
then always emits the flag:

```objc
if (sshPort) {
    TA(@"-p", ([NSString stringWithFormat:@"%ld", (long)sshPort]));
}
```

The `if (sshPort)` guard was written for the previous call site in `SPConnectionController.m`,
which passed `[[self sshPort] integerValue]` — `0` for an empty field — so no `-p` was added and
ssh fell back to the config's `Port`:

```objc
sshTunnel = [[SPSSHTunnel alloc] initToHost:[self sshHost] port:[[self sshPort] integerValue] ...];
```

That call site was removed in e40e67f, and the `?? 22` default came in with b001c1f. Both landed
via **PR #2375 ("Modernization: Decouple connection flow, extract to Swift"), merged 2026-04-17**.
The Swift connection service is not present at the 5.2.1 (20100) tag (released 2026-04-14, before
#2375 merged) and is present at 5.3.1 (20104), so this **first shipped in the 5.3.x line**.

**Suggested Fix**

Preserve the sentinel the guard expects:

```swift
let sshPort = info.sshPort.isEmpty ? 0 : (Int(info.sshPort) ?? 0)
```

**Workaround**

Type the config's `Port` value into the SSH Port field. The alias can stay in SSH Host —
`HostName` and `User` still resolve from `~/.ssh/config`.

**Related Issues**

- #2385 and #2491 — other regressions traced to the same PR #2375 Swift rewrite (#2491 is
  explicitly described as "introduced in version 5.3"), which corroborates 5.3.x as the point of
  introduction for this class of SSH/connection regressions.

### PR DESCRIPTION
**Pull request by Jason-Morcos**
## Changes:
- Preserve the SSH tunnel API's zero-port sentinel when the SSH Port field is empty, allowing OpenSSH to use the selected config file's Port directive.
- Accept explicit SSH ports from 1 through 65535 as command-line overrides, and reject malformed or out-of-range values before tunnel creation.
- Add pure Swift regression coverage for blank, valid, and invalid SSH port mappings.

## Closes following issues:
- Closes: #2513

## Tested:
- Processors:
  - [ ] Intel
  - [x] Apple Silicon
- macOS Versions:
  - [ ] 12.x (Monterey)
  - [ ] 13.x (Ventura)
  - [ ] 14.x (Sonoma)
  - [ ] 15.x (Sequoia)
  - [x] 26.x
- Localizations:
  - [x] English
  - [ ] Spanish
  - [ ] Other (please specify)
- Xcode Version:
  - 26.6 (17F113)
- Focused SAConnectionInfoMappingTests: 26 passed, 0 failed
- Full Unit Tests scheme: 790 passed, 60 skipped, 0 failed
- Sequel Ace Debug build: passed
- OpenSSH config-resolution check: config Port 2222 is retained without -p and overridden when -p 22 is supplied

## Screenshots:
- Not applicable; no visual changes.

## Additional notes:
- SPSSHTunnel already omits the -p argument when initialized with port 0. When neither the field nor SSH config sets a port, OpenSSH still defaults to port 22. This restores the sentinel used by the legacy connection path before the Swift service extraction.
- A live end-to-end SSH/MySQL connection was not run because no issue-specific remote test host was available.

## Summary by CodeRabbit

* **New Features**
  * Added support for configuring a custom SSH port for connections.
  * Port values are trimmed and validated against the allowed range of 1–65535.
  * Blank SSH port values defer to SSH configuration; invalid explicit values display a localized validation error.

* **Tests**
  * Added coverage for blank, whitespace-only, valid, and invalid SSH port values.

### CODE REVIEW

**Comment by Jason-Morcos** [comment]
@codex review

File: Source/Model/SAConnectionInfo.swift [hunk 1]
```diff
@@ -357,6 +357,11 @@ struct SAConnectionInfo {
         set { info.sshPort = newValue }
     }
 
+    /// The explicit command-line port override, or zero to defer to the SSH configuration.
+    var sshPortOverride: Int {
+        Int(info.sshPort) ?? 0
+    }
```
[the remark this answers is not in the document]
Comment (Jason-Morcos) [reply Source/Model/SAConnectionInfo.swift]: Fixed in 65ac6c4b9. SAConnectionInfo.sshPortOverride now trims the field, maps only a blank value to 0, accepts ports 1...65535, and returns nil for malformed or out-of-range input. SAConnectionService reports that validation failure before constructing SPSSHTunnel. Tests cover blank and whitespace input, boundary values, malformed input, zero, negative, and out-of-range values.

File: UnitTests/SAConnectionServiceTests.swift [hunk 2]
```diff
@@ -62,6 +62,20 @@ final class SAConnectionInfoMappingTests: XCTestCase {
...
+        info.sshPort = ""
+
+        XCTAssertEqual(info.sshPortOverride, 0)
+    }
+
+    func testExplicitSSHPortOverridesSSHConfiguration() {
+        let info = SAConnectionInfoObjC()
+        info.sshPort = "2222"
+
+        XCTAssertEqual(info.sshPortOverride, 2222)
```
[the remark this answers is not in the document]
Comment (Jason-Morcos) [reply UnitTests/SAConnectionServiceTests.swift]: Addressed in 65ac6c4b9. SAConnectionInfoObjC is a Swift @objc wrapper rather than a legacy Objective-C project type, and the existing mapping suite already exercises that wrapper. Still, the new port mapping now lives on pure Swift SAConnectionInfo, and the added port tests exercise that value type directly, as requested.

**Comment by Jason-Morcos** [comment]
@codex review

File: Source/Model/SAConnectionInfo.swift [hunk 3]
```diff
@@ -89,6 +89,16 @@ struct SAConnectionInfo {
...
     var sshKeyLocation: String = ""
     var sshPort: String = ""
+
+    /// The explicit command-line port override. A blank value returns zero so
+    /// OpenSSH can resolve the port from its configuration; invalid values fail.
+    var sshPortOverride: Int? {
+        let trimmedPort = sshPort.trimmingCharacters(in: .whitespacesAndNewlines)
+        guard !trimmedPort.isEmpty else { return 0 }
+        guard let port = Int(trimmedPort), (1...65535).contains(port) else { return nil }
+        return port
```
[the remark this answers is not in the document]
Comment (Jason-Morcos) [reply Source/Model/SAConnectionInfo.swift]: Fixed in a5218be7e. Both guard returns are now on their own lines, matching conditional_returns_on_newline without changing the mapping behavior. The 26 focused mapping tests still pass.

**Comment by Jason-Morcos** [comment]
@codex review

**Comment by scriptease** [comment]
Verified this fixes #2513. Root cause matches: \`Int(info.sshPort) ?? 22\` defaulted a blank SSH Port field to 22, clobbering the \`Port\` directive resolved from \`~/.ssh/config\`. Confirmed \`SPSSHTunnel.m\`'s \`if (sshPort) { TA(@"-p", ...) }\` already omits \`-p\` when the port is 0, so restoring the zero-sentinel via \`sshPortOverride\` correctly lets OpenSSH fall back to the config file's Port. Also verified the added validation (reject 0/negative/>65535/non-numeric) and the new unit tests cover blank, valid, and invalid cases correctly.

**Comment by Jason-Morcos** [comment]
@Kaspik Could you take a quick look at this one when you get the chance? I believe this was a regression in the big 5.3 refactors
